platforms/netbsd/hax_wrapper.c

```diff
@@ -251,36 +251,34 @@ hax_atomic_t hax_atomic_dec(volatile hax_atomic_t *atom)
 
 int hax_test_and_set_bit(int bit, uint64_t *memory)
 {
-    volatile uint64_t *val;
-    uint64_t mask, old;
-
-    val = (volatile uint64_t *)memory;
-    mask = 1 << bit;
+    const unsigned int units = (sizeof(*memory) * CHAR_BIT);
+    volatile unsigned long *const p = &memory[bit / units];
+    const unsigned long mask = (1UL << (bit % units));
+    unsigned long v;
 
+    membar_exit();
```

Comment:
nit: Could you enlighten me on the use of memory barriers here? Is the atomic operation (CAS) not enough to guarantee the correct behavior?

Reply:
I've followed Linux that implies memory barrier.

```
/**
 * test_and_set_bit - Set a bit and return its old value
 * @nr: Bit to set
 * @addr: Address to count from
 *
 * This operation is atomic and cannot be reordered.
 * It also implies a memory barrier.
 */
static __always_inline bool test_and_set_bit(long nr, volatile unsigned long *addr)
{
	return GEN_BINARY_RMWcc(LOCK_PREFIX __ASM_SIZE(bts), *addr, c, "Ir", nr);
}
```

Reply:
I see. I guess the underlying x86 implementation for `atomic_cas_ulong()` already implies a memory barrier, but since it's not specified in the documentation, your approach is more portable.